pxr/base/tf/delegatedCountPtr.h

```diff
@@ -0,0 +1,271 @@
+    /// Assigns by copying a convertible pointer from `ptr`.
+    ///
+    /// This is necessary in part to allow a `const` `ValueType` to be
+    /// assigned from from a non-`const` `ValueType`.
+    ///
+    /// Retains only if the pointer is valid.
+    template <typename OtherType>
+    TfDelegatedCountPtr& operator=(
+        const TfDelegatedCountPtr<OtherType>& ptr)
+    noexcept(IncrementAndDecrementAreNoExcept()) {
+        static_assert(_IsPtrConvertible<OtherType>::value);
+        // Implement copy assigment in terms of move assignment
+        return (*this = TfDelegatedCountPtr{ptr});
+    }
+
+    /// Assigns by moving the pointer from `ptr`.
+    ///
+    /// `ptr` will be reset to its default state (ie. `nullptr`).
+    TfDelegatedCountPtr& operator=(TfDelegatedCountPtr&& ptr)
+    noexcept(DecrementIsNoExcept()) {
+        _DecrementIfValid();
+        _pointer = ptr.get();
+        ptr._pointer = nullptr;
+        return *this;
+    }
+
+    /// Resets the pointer to its default state (ie. `nullptr`)
+    TfDelegatedCountPtr& operator=(std::nullptr_t)
+    noexcept(DecrementIsNoExcept()) {
+        reset();
+        return *this;
+    }
+
+    /// Releases if the underlying pointer is valid 
+    ~TfDelegatedCountPtr() noexcept(DecrementIsNoExcept()) {
+        _DecrementIfValid();
+    }
+
+    /// Dereference the underlying pointer
+    ReferenceType operator*() const { 
+        return *get();
+    }
+
+    /// Arrow operator dispatch for the underlying pointer
+    RawPtrType operator->() const noexcept {
+        return get();
+    }
+
+    /// Returns true if the underlying pointer is non-null.
+    explicit operator bool() const noexcept { return get(); }
+
+    /// Returns true if the underlying pointers are equivalent.
+    template <typename OtherType>
+    bool operator==(
+        const TfDelegatedCountPtr<OtherType>& other) const noexcept {
+        return get() == other.get();
+    }
+
+    /// Returns false if the underlying pointers are equivalent.
+    template <typename OtherType>
+    bool operator!=(
+        const TfDelegatedCountPtr<OtherType>& other) const noexcept {
+        return get() != other.get();
+    }
+
+    /// Orders based on the underlying pointer.
+    template <typename OtherType>
+    bool operator<(
+        const TfDelegatedCountPtr<OtherType>& other) const noexcept {
+        return get() < other.get();
+    }
+
+    /// Return the underlying pointer.
+    RawPtrType get() const noexcept { return _pointer; }
+
+    /// Reset the pointer to its default state (`nullptr`), releasing if
+    /// necessary.
+    void reset() noexcept(DecrementIsNoExcept()) {
+        _DecrementIfValid();
+        _pointer = nullptr;
+    }
+
+    /// Swaps the pointer between two delegated count pointers
+    void swap(TfDelegatedCountPtr& other) noexcept {
+        std::swap(other._pointer, _pointer);
+    }
+
+private:
+    void _IncrementIfValid() noexcept(IncrementIsNoExcept()) {
+        if (get()) {
```

Comment:
any reason not to just do `if (_pointer)` here?  we use it directly in the body immediately after -- it seems clearer to me that way.

Reply:
A bunch of prior usages of `_pointer` had to be converted to `get()` to support operations on convertible types without `friend`-ship.  But there's no reason this usage had to be converted.